src/observer/CameraSensorServer.cpp

```diff
@@ -0,0 +1,439 @@
+
+    // Get depth frame
+    auto frame = frames.get_depth_frame();
+    frame = frame.apply_filter(dec);
+    frame = frame.apply_filter(hdr);
+    frame = frame.apply_filter(thresh);
+    frame = frame.apply_filter(depth2disparity);
+    frame = frame.apply_filter(spat);
+    frame = frame.apply_filter(disparity2depth);
+
+    rs2_intrinsics intrinsics = frame.get_profile().as<rs2::video_stream_profile>().get_intrinsics();
+
+    const size_t height = static_cast<size_t>(frame.get_height());
+    const size_t width = static_cast<size_t>(frame.get_width());
+
+    const size_t half_height = static_cast<size_t>(static_cast<double>(height) * 0.5);
+    const size_t half_width = static_cast<size_t>(static_cast<double>(width) * 0.5);
+
+    std::array<float, 2> pixel;
+    const size_t half_kernel_size = kernel_size_ / 2;
+    std::vector<Eigen::Vector3d> points;
+    const size_t offset = 110;
+    for(size_t i = half_kernel_size - offset; i < width - half_kernel_size - offset; ++i)
+    {
+      pixel[0] = static_cast<float>(i);
+      pixel[1] = static_cast<float>(half_height);
+
+      const float depth = applyKernel(pixel, frame);
+      const Eigen::Vector3d point = depthToPoint(intrinsics, pixel, depth);
+      if(point.z() != 0)
+      {
+        points.push_back(point);
+      }
+    }
+
+    // Sort
+    std::sort(points.begin(), points.end(),
+              [](const Eigen::Vector3d & a, const Eigen::Vector3d & b) { return a.x() < b.x(); });
+
+    {
+      const std::lock_guard<std::mutex> lock(points_mtx_);
+      points_ = points;
+    }
+    if(data_->client.isAlive())
+    {
+      data_->data_ready->notify_all();
+    }
+  }
+}
+
+void CameraSensorServer::computation()
+{
+  pc_estimated_ground_points_ = std::make_shared<open3d::geometry::PointCloud>();
+  pc_transformed_estimated_ground_points_ = std::make_shared<open3d::geometry::PointCloud>();
+  pc_full_ground_reconstructed_points_ = std::make_shared<open3d::geometry::PointCloud>();
+  bool client_was_alive = data_->client.isAlive();
+  while(run_)
+  {
+    while(data_->client.isAlive())
+    {
+      if(!client_was_alive)
+      {
+        mc_rtc::log::success("[CameraSensorServer::{}] Client connected", name_);
+        client_was_alive = true;
+      }
+      ipc::scoped_lock<ipc::interprocess_mutex> lck(data_->mtx);
+      if(data_->compute_ready->timed_wait(lck, pclock::universal_time() + ptime::seconds(1)))
+      {
+        do_computation();
+      }
+    }
+    if(client_was_alive)
+    {
+      client_was_alive = false;
+      mc_rtc::log::critical("[CameraSensorServer::{}] Client disconnected", name_);
+    }
+    // Wait for the client to become available
+    data_->server_idle.notify_all();
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+}
+
+void CameraSensorServer::do_computation()
+{
+  if(data_->skip)
+  {
+    return;
+  }
+  {
+    const std::lock_guard<std::mutex> lock(points_mtx_);
+    ground_points_ = points_;
```

Comment:
A note here: this is very different from the current master, on the current master `ground_points_` are used here but they are never filled up, from reading the code I think this is the correct intent?